ta/gatekeeper/gatekeeper_ta.c

```diff
@@ -0,0 +1,767 @@
+
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <utee_defines.h>
+
+#include "failure_record.h"
+#include "gatekeeper_ipc.h"
+#include "ta_gatekeeper.h"
+
+static const uint8_t secret_id[] = {0xB1, 0x6B, 0x00, 0xB5};
+
+TEE_Result TA_CreateEntryPoint(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+	TEE_ObjectHandle secret_obj = TEE_HANDLE_NULL;
+	uint8_t secret_data[TEE_SHA256_HASH_SIZE];
+
+	DMSG("Checking master key secret");
+	res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE, secret_id,
+				       sizeof(secret_id),
+				       TEE_DATA_FLAG_ACCESS_READ,
+				       &secret_obj);
+
+	if (!res) {
+		DMSG("Secret is already created");
+		goto exit;
+	}
+
+	if (res != TEE_ERROR_ITEM_NOT_FOUND) {
+		EMSG("Failed to open secret, error=%X", res);
+		return res;
+	}
+
+	DMSG("Create master key secret");
+
+	TEE_GenerateRandom(secret_data, sizeof(secret_data));
+
+	res = TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE,
+			secret_id,
+			sizeof(secret_id),
+			TEE_DATA_FLAG_ACCESS_WRITE,
+			TEE_HANDLE_NULL, NULL, 0,
+			&secret_obj);
+	if (res) {
+		EMSG("Failed to create secret");
+		return res;
+	}
+
+	res = TEE_WriteObjectData(secret_obj, (void *) secret_data,
+			sizeof(secret_data));
+	if (res)
+		EMSG("Failed to write secret data");
+
+exit:
+	if (secret_obj != TEE_HANDLE_NULL)
+		TEE_CloseObject(secret_obj);
+
+	return res;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS] __unused,
+				    void **sess_ctx __unused)
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	init_failure_record();
+
+	return TEE_SUCCESS;
+}
+
+void TA_CloseSessionEntryPoint(void *sess_ctx __unused)
+{
+}
+
+static TEE_Result TA_GetMasterKey(TEE_ObjectHandle master_key)
+{
+	TEE_Result res;
+	TEE_Attribute attr;
+	uint8_t	secret_data[TEE_SHA256_HASH_SIZE];
+	TEE_ObjectHandle secret_obj = TEE_HANDLE_NULL;
+	uint32_t read_size = 0;
+
+	res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE, secret_id,
+				       sizeof(secret_id),
+				       TEE_DATA_FLAG_ACCESS_READ,
+				       &secret_obj);
+	if (res) {
+		EMSG("Failed to open secret, error=%X", res);
+		goto exit;
+	}
+
+	res = TEE_ReadObjectData(secret_obj, secret_data, sizeof(secret_data),
+				 &read_size);
+	if (res || sizeof(secret_data) != read_size) {
+		EMSG("Failed to read secret data, bytes = %u", read_size);
+		goto close_obj;
+	}
+
+	TEE_InitRefAttribute(&attr, TEE_ATTR_SECRET_VALUE, secret_data,
+			     sizeof(secret_data));
+
+	res = TEE_PopulateTransientObject(master_key, &attr, 1);
+	if (res)
+		EMSG("Failed to set master key attributes");
+
+close_obj:
+	TEE_CloseObject(secret_obj);
+exit:
+	return res;
+}
+
+static TEE_Result TA_ComputeSignature(uint8_t *signature,
```

Comment:
No strong opinion here, but this is more or less a plain HMAC operation. Might be worth adding that as a local function with just in/out parameters and just call that from here. As an example, see [this](https://github.com/linaro-swg/optee_examples/blob/master/hotp/ta/hotp_ta.c#L33-L112) function and how that is used. Another example is when I taking my first stab at doing a Gatekeeper implementation, I ended up with something like [this](https://github.com/jbech-linaro/optee_examples/blob/optee_gatekeeper_v0.1/gatekeeper/ta/gatekeeper_ta.c#L142-L165).